src/pk.c

```diff
@@ -51,6 +51,487 @@ static int pem_write_bio_pubkey(WOLFSSL_BIO* bio, WOLFSSL_EVP_PKEY* key);
+ *
+ * @param [in, out] bio         BIO object to read with.
+ * @param [out]     pem         Read data in a buffer.
+ * @param [out]     pemSz       Amount of data read in bytes.
+ * @param [out]     memAlloced  Indicates whether return buffer was allocated.
+ * @return  -ve on error.
+ * @return  0 on success.
+ */
+static int wolfssl_read_bio(WOLFSSL_BIO* bio, char** pem, int* pemSz,
+    int* memAlloced)
+{
+    int ret;
+    int sz;
+
+    if (bio->type == WOLFSSL_BIO_MEMORY) {
+        ret = wolfSSL_BIO_get_mem_data(bio, pem);
+        if (ret > 0) {
+            bio->rdIdx += ret;
+        }
+        *memAlloced = 0;
+    }
+#ifndef WOLFSSL_NO_FSEEK
+    /* Get pending or, when a file BIO, get length of file. */
+    else if ((sz = wolfSSL_BIO_get_len(bio)) > 0) {
+        ret = wolfssl_read_bio_len(bio, sz, pem);
+        if (ret > 0) {
+            *memAlloced = 1;
+        }
+    }
+#else
+    else if ((sz = wolfSSL_BIO_pending(bio)) > 0) {
+        ret = wolfssl_read_bio_len(bio, sz, pem);
+        if (ret > 0) {
+            *memAlloced = 1;
+        }
+    }
+    else if (bio->type == WOLFSSL_BIO_FILE) {
+        ret = wolfssl_read_bio_file(bio, pem);
+        if (ret > 0) {
+            *memAlloced = 1;
+        }
+    }
+#endif
+    else {
+        WOLFSSL_MSG("No data to read from bio");
+        *memAlloced = 0;
+        ret = NOT_COMPILED_IN;
+    }
+
+    if (ret >= 0) {
+        *pemSz = ret;
+        ret = 0;
+    }
+
+    return ret;
+}
+#endif /* !NO_BIO */
+
+#if !defined(NO_FILESYSTEM)
+/* Read all the data from a file.
+ *
+ * @param [in]  fp          File pointer to read with.
+ * @param [out] pem         Read data in an allocated buffer.
+ * @param [out] pemSz       Amount of data read in bytes.
+ * @return  WOLFSSL_BAD_FILE on error.
+ * @return  0 on success.
+ */
+static int wolfssl_read_file(XFILE fp, char** pem, int* pemSz)
+{
+    int ret = 0;
+    int sz = 0;
+    char* mem = NULL;
+
+    if (fp == XBADFILE) {
+        ret = WOLFSSL_BAD_FILE;
+    }
+    if (ret == 0) {
+        /* Read whole file into a new buffer. */
+        if (XFSEEK(fp, 0, SEEK_END) != 0) {
+            ret = WOLFSSL_BAD_FILE;
+        }
+    }
+    if (ret == 0) {
+        sz = (int)XFTELL(fp);
+        if (sz < 0) {
+            ret = WOLFSSL_BAD_FILE;
+        }
+    }
+    if ((ret == 0) && (XFSEEK(fp, 0, SEEK_SET) != 0)) {
+        ret = WOLFSSL_BAD_FILE;
+    }
+    if ((ret == 0) && ((sz > MAX_WOLFSSL_FILE_SIZE) || (sz <= 0L))) {
+        ret = WOLFSSL_BAD_FILE;
+    }
+    if (ret == 0) {
+        mem = (char*)XMALLOC(sz, NULL, DYNAMIC_TYPE_PEM);
+        if (mem == NULL) {
+            ret = WOLFSSL_BAD_FILE;
+        }
+    }
+    if ((ret == 0) && ((int)XFREAD(mem, 1, sz, fp) != sz)) {
+        ret = WOLFSSL_BAD_FILE;
+    }
+    if (ret == 0) {
+        *pemSz = sz;
+        *pem = mem;
+        mem = NULL;
+    }
+
+    XFREE(mem, NULL, DYNAMIC_TYPE_PEM);
+    return ret;
+}
+#endif /* !NO_FILESYSTEM */
+#endif
```

Comment:
A comment on this \#endif would be helpful.

Reply:
Done.